include/devicetree/reserved.h

```diff
@@ -0,0 +1,137 @@
+/**
+ * @file
+ * @brief reserved-memory Devicetree macro public API header file.
+ */
+
+/*
+ * Copyright (c) 2021 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DEVICETREE_RESERVED_H_
+#define ZEPHYR_INCLUDE_DEVICETREE_RESERVED_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup devicetree-reserved Devicetree reserved-memory API
+ * @ingroup devicetree
+ * @{
+ */
+
+#include <linker/devicetree_reserved.h>
+
+/**
+ * @brief Get the pointer to the reserved-memory region
+ *
+ * Example devicetree fragment:
+ *
+ *     reserved: reserved-memory {
+ *         compatible = "reserved-memory";
+ *         ...
+ *         n: node {
+ *             reg = <0x42000000 0x1000>;
+ *         };
+ *      };
+ *
+ * Example usage:
+ *
+ *     DT_RESERVED_GET_ADDR(DT_NODELABEL(n)) // (uint8_t *) 0x42000000
+ *
+ * @param node_id node identifier
+ * @return pointer to the beginning of the reserved-memory region
+ */
+#define DT_RESERVED_GET_ADDR(node_id) _DT_RESERVED_START(node_id)
+
+/**
+ * @brief Get the size of the reserved-memory region
+ *
+ * Example devicetree fragment:
+ *
+ *     reserved: reserved-memory {
+ *         compatible = "reserved-memory";
+ *         ...
+ *         n: node {
+ *             reg = <0x42000000 0x1000>;
+ *         };
+ *     };
+ *
+ * Example usage:
+ *
+ *     DT_RESERVED_GET_SIZE(DT_NODELABEL(n)) // 0x1000
+ *
+ * @param node_id node identifier
+ * @return the size of the reserved-memory region
+ */
+#define DT_RESERVED_GET_SIZE(node_id) DT_REG_SIZE(node_id)
+
+/**
+ * @brief Get the pointer to the reserved-memory region from a memory-reserved
+ *        phandle
+ *
+ * Example devicetree fragment:
+ *
+ *     reserved: reserved-memory {
+ *         compatible = "reserved-memory";
+ *         ...
+ *         res0: res {
+ *             reg = <0x42000000 0x1000>;
+ *             label = "res0";
+ *         };
+ *     };
+ *
+ *     n: node {
+ *         memory-region = <&res0>;
+ *     };
+ *
+ * Example usage:
+ *
+ *     DT_RESERVED_GET_ADDR_BY_PHANDLE(DT_NODELABEL(n), memory_region) // (uint8_t *) 0x42000000
+ *
+ * @param node_id node identifier
+ * @return pointer to the beginning of the reserved-memory region
+ */
+#define DT_RESERVED_GET_ADDR_BY_PHANDLE(node_id, res_pha) \
+	DT_RESERVED_GET_ADDR(DT_PHANDLE_BY_IDX(node_id, res_pha, 0))
```

Comment:
- Can just use DT_PHANDLE instead of DT_PHANDLE_BY_IDX
- I have similar reservations about using ADDR in the name here